### ISSUE TITLE
MiniBatch Training : encoder-decoder-coupling.lua

### ISSUE THREAD

**Issue by zhongzho** [Element-Research/rnn#142]
Hi Nicholas, 

I am trying to use encoder-decoder-coupling.lua for mini-batch training. May you share with us a sample of how to modify the layers to take care of the extra dimension of batchSize? 

Thank you! 

Cheers, 
Zhong

**Comment by cjmcmurtrie** [comment]
I have been trying something similar for several weeks. About three times I thought I was there, then realised I wasn't. Help would be useful.

**Comment by nicholas-leonard** [comment]
@cjmcmurtrie @zhongzho It already does. I modified the example to make this more explicit : https://github.com/Element-Research/rnn/commit/22b038d43733ed41c7f59277ce7f71d24b74a970

**Comment by cjmcmurtrie** [comment]
What shape/type should the mini-batches be?

I have been trying it with zero-padded tensors of variable length sequences as rows (using LookupTableMaskZero etc) and it doesn't work so far.

**Comment by nicholas-leonard** [comment]
@cjmcmurtrie I added a unit test to see if it handles variable batch sizes : https://github.com/Element-Research/rnn/commit/e8940edad48dc399f4a2e8f4b399cb0dfed1cd5b . I also simplified the example to include a training loop. Anyway, if you are still having issues, trying making a minimal modification of the unit test to make it fail.

**Comment by cjmcmurtrie** [comment]
Hi Nicholas, here's the problem. When you use the setup in the example, with batch size = 1, the model can encode and decode a sequence. An example decoding after training with 1000 sentences for 2 epochs:
```
the name help whewell provided circle : eventually learned science fiction mathematician rooms this definition magazines eventually learned science
```
However, it no longer works when you make the batch size larger than 1. The training error decreases unsuspiciously, but when you actually try to use the model to generate a decoded sequences, this is what it produces (repeating the end of sentence token):
```
EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS EEOOSS
```
What do you think?

**Comment by nicholas-leonard** [comment]
@cjmcmurtrie In the example the batch size is 2 : 
```lua
local encInSeq = torch.Tensor({{1,2,3},{3,2,1}}) 
local decInSeq = torch.Tensor({{1,2,3,4},{4,3,2,1}})
local decOutSeq = torch.Tensor({{2,3,4,1},{1,2,4,3}})
decOutSeq = nn.SplitTable(1, 1):forward(decOutSeq)
```
So for the `encInSeq` the first sequence is `{1,2,3}`, the second is `{3,2,1}`.

You seem to be using a different example though, so the error may lie in your own code, which I don't have access too :)

**Comment by cjmcmurtrie** [comment]
Hey Leonard. I've reproduced a minimal code example to show you what I mean, see below. Notice that the toy data batch is zero-padded, and I have used MaskZero, LookupTableMaskZero and MaskZeroCriterion. I am using 1 as a start sequence token, and 100 as an end sequence token.

If you run the script below, you will notice that the output() function produces empty tensors when generating a sequence. This is because the first prediction is always the end token (100).

However, this behaviour is not fixed. If you change the line that calls this function to a different input sequence, for instance to:
```
print (output(torch.CudaTensor{71, 45, 87}, model, 10))
```
Then after some iterations of generating empty sequences, it does start to generate one of the training set sequences:
```
Loss:2.2462656497955	
Loss:2.176420211792	
[torch.DoubleTensor with no dimension]

Loss:2.113420009613	
Loss:2.0567095279694	
[torch.DoubleTensor with no dimension]

Loss:2.0055947303772	
Loss:1.9593524932861	
 39
 40
 62
 63
[torch.DoubleTensor of size 4]
```
Let me know your thoughts!
```
require 'nn'
require 'rnn'
require 'optim'
require 'cutorch'
require 'cunn'



local vector_dim = 20
local hidden = 40
local vocab_size = 100
local max_len = 6


torch.manualSeed(123)

local model = {}

model.encoderLSTM = nn.FastLSTM(vector_dim, hidden)
model.decoderLSTM = nn.FastLSTM(vector_dim, hidden)

model.encoder = nn.MaskZero(nn.Sequential()
				:add(nn.LookupTableMaskZero(vocab_size, vector_dim))
				:add(nn.SplitTable(1, 2))
				:add(nn.Sequencer(model.encoderLSTM))
				:add(nn.SelectTable(-1)), 1):cuda()
model.encoder:zeroGradParameters()

model.decoder = nn.MaskZero(nn.Sequential()
					:add(nn.LookupTableMaskZero(vocab_size, vector_dim))
					:add(nn.SplitTable(1, 2))
					:add(nn.Sequencer(model.decoderLSTM))
					:add(nn.Sequencer(nn.Linear(hidden, vocab_size)))
					:add(nn.Sequencer(nn.LogSoftMax())), 1):cuda()

model.decoder:zeroGradParameters()

model.zerotensor = torch.Tensor(2):zero():cuda()
model.optimstate = {learningRate = 0.005, alpha = 0.95}
model.criterion = nn.SequencerCriterion(nn.MaskZeroCriterion(nn.ClassNLLCriterion(), 1)):cuda()

local encoder_input = torch.CudaTensor{{0, 0, 39, 40, 62, 63}, {0, 0, 0, 0, 98, 97}}
local decoder_input = torch.CudaTensor{{0, 1, 39, 40, 62, 63}, {0, 0, 0, 21, 98, 97}}
local decoder_target = torch.CudaTensor{{0, 39, 40, 62, 63, 100}, {0, 0, 0, 98, 97, 100}}




function forwardConnect(model, seq_len)

  model.decoderLSTM.userPrevOutput = 
  	nn.rnn.recursiveCopy(model.decoderLSTM.userPrevOutput, model.encoderLSTM.outputs[seq_len])
  model.decoderLSTM.userPrevCell = 
  	nn.rnn.recursiveCopy(model.decoderLSTM.userPrevCell, model.encoderLSTM.cells[seq_len])

end

function backwardConnect(model)

  model.encoderLSTM.userNextGradCell = 
  	nn.rnn.recursiveCopy(model.encoderLSTM.userNextGradCell, model.decoderLSTM.userGradPrevCell)
  model.encoderLSTM.gradPrevOutput = 
  	nn.rnn.recursiveCopy(model.encoderLSTM.gradPrevOutput, model.decoderLSTM.userGradPrevOutput)

end




-- Return a prediction for an input sequence
function output(input, model, max_out)

	model.encoderLSTM:remember()
	model.decoderLSTM:remember()

	model.encoder:forward(input)
	forwardConnect(model, input:size(1))

	local sequence = {}

	-- -- Forward <go> and all of it's output recursively back to the decoder
	local output = torch.CudaTensor{1}

	for i = 1, max_out do

		local prediction = model.decoder:forward(output)[1]

		-- prediction contains the probabilities for each word IDs.
		-- The index of the probability is the word ID.
		local prob, preds = prediction:sort(1, true)

		-- First one is the most likely.
		output = torch.CudaTensor{preds[1]}

		-- Terminate on EOS token
		if output[1] == 100 then
			break
		end

		table.insert(sequence, preds[1])
	end 

	model.encoder:forget()
	model.decoder:forget()

	return torch.Tensor(sequence)

end






function feval(x)

	if x ~= params then
        params:copy(x)
    end
    gradParams:zero()

    decoder_target_split = nn.SplitTable(1, 1):forward(decoder_target)

    -- Forward pass
    model.encoder:forward(encoder_input)
    forwardConnect(model, max_len)
    local decoder_output = model.decoder:forward(decoder_input)

    local loss = model.criterion:forward(decoder_output, decoder_target_split)

    -- Backward pass
    gEdec = model.criterion:backward(decoder_output, decoder_target_split)
	model.decoder:backward(decoder_input, gEdec)
	backwardConnect(model)
	model.encoder:backward(encoder_input, model.zerotensor)

	model.encoder:gradParamClip(2)
	model.decoder:gradParamClip(2)

	collectgarbage()

	return loss, gradParams

end




modules = nn.Parallel()
    :add(model.encoder)
    :add(model.decoder)

params, gradParams = modules:getParameters()

local epochs = 1000

for e = 1 , epochs do
	_ , loss = optim.adam(feval, params, model.optimstate)
	print ('Loss:' .. tostring(loss[1]))
	if e % 2 == 0 then
		print (output(torch.CudaTensor{39, 40, 62}, model, 10))
	end
end
```

**Comment by cjmcmurtrie** [comment]
What seems to be happening is that the model over-estimates the probability of the end-of-sequence token, predicting it immediately after every instance of the start of sequence token. This only seems to happen when zero padding is used for batch_size > 1 (zero padding is fine for batch_size = 1).

In support of this, if I choose the **second most probable word** greedily, rather than the first (which is always the end of sequence token), then after some epochs, (somewhat) sensible sequences begin to appear, e.g.:
```
it they there
some people call : what is beauty ? :
```
It is not yet making sense in my mind why the model would put so much weight on the end-of-sequence token as a result of variable length sequences in the batch.

**Comment by htw2012** [comment]
@cjmcmurtrie @nicholas-leonard  I think it is not the model would put so much weight on the end of sequence token.Just your training is not full, you may get more iterations in your model.I use code to do the task of translation I also get the same problem.
By the way,could you show me the model of test precedures? I can't get decoder output to feed the decoder input at each time,sorry for my torch.

**Comment by rahulgupta-iisc** [comment]
@cjmcmurtrie Were you able to solve your issue?
